native/lib.rs

```diff
@@ -58,29 +60,32 @@ fn put_5<'a>(
     max_size: usize,
     min_q1_size: usize,
 ) -> Token {
-    let mut caches = CACHES.lock().unwrap();
     let cache_name_bin = HBin::from_atom(env, cache_name);
-    let cache = match caches.get_mut(&cache_name_bin) {
-        Some(cache) => cache,
-        None => {
-            let recent_ratio = min_q1_size as f64 / max_size as f64;
-            let inner = TwoQueueCacheBuilder::new(max_size)
-                .set_recent_ratio(recent_ratio)
-                .finalize()
-                .map_err(|e| {
-                    eprintln!(
-                        "MinQ1Size/MaxSize ({}/{}) ratio doesn't appear to make sense ({})",
-                        min_q1_size, max_size, e
-                    )
-                })
-                .expect("MinQ1Size/MaxSize ratio doesn't appear to make sense");
-            let cache = TermCache::new(inner);
-            caches.insert(HBin::from_atom(env, cache_name), cache);
-            caches
-                .get_mut(&cache_name_bin)
-                .expect("cache was just inserted. It must exist")
+    let cache = {
+        match CACHES.read().unwrap().get(&cache_name_bin) {
+            Some(cache) => cache.clone(),
+            None => {
+                let recent_ratio = min_q1_size as f64 / max_size as f64;
+                let inner = TwoQueueCacheBuilder::new(max_size)
+                    .set_recent_ratio(recent_ratio)
+                    .finalize()
+                    .map_err(|e| {
+                        eprintln!(
+                            "MinQ1Size/MaxSize ({}/{}) ratio doesn't appear to make sense ({})",
+                            min_q1_size, max_size, e
+                        )
+                    })
+                    .expect("MinQ1Size/MaxSize ratio doesn't appear to make sense");
+                let cache = Arc::new(RwLock::new(TermCache::new(inner)));
+                CACHES
+                    .write()
```

Comment:
Is this trying to get a `write` lock on `CACHES` without releasing the `read` lock?

Reply:
This code looks like it's creating a new cache for `put` if one doesn't exist. The code here differs slightly from the code for `create`. Can the cache craeting get moved to a utility function that both `create` and `put` use?

Reply:
> Can the cache craeting get moved to a utility function that both create and put use?

Probably. _Maybe_ there's a valid reason I didn't do that, but it's likely an oversight.

Reply:
> It may not be obvious, but the read lock acquired on is a temporary and dropped at the end of line 65

Got it. The lock gets dropped when the value goes out of scope. This is my non-knowledge of Rust getting in the way...

Reply:
Follow up: Marc's intuition is correct and there is in fact a deadlock here. It is fixed with a recent commit.